File: frontend/src/features/plantillas/instrucciones.ts
import { hayBackend, pedirAlBackend } from '@/shared/api/backend'
import type { ResultadoDeConsulta } from '@/shared/supabase/consultas'
import type { MarcadorDeDocx, MarcadorSimpleDeDocx } from './data'
import { nombreDeMarcador } from './plantillas'

/*
  Pedirle a la IA qué debe escribirse en cada campo de la plantilla.

  Configurar quince campos a mano es donde se abandona una plantilla, y el
  documento ya trae casi todo: cómo se llama el campo y qué hay escrito a su
  alrededor. Esto los manda al backend (`POST /plantillas/instrucciones`) y
  devuelve la configuración propuesta.

  **No pisa lo ya escrito.** Solo se ofrece rellenar los campos que están sin
  instrucción: quien configuró uno a mano tenía una razón.
*/

export type CampoPropuesto = {
  readonly id: string
  readonly instruccion: string
  readonly formato: MarcadorSimpleDeDocx['formato']
  readonly modo: NonNullable<MarcadorSimpleDeDocx['modo']>
  readonly extension: NonNullable<MarcadorSimpleDeDocx['extension']>
}

export function camposSinInstruccion(marcadores: readonly MarcadorDeDocx[]): readonly MarcadorSimpleDeDocx[] {
  return marcadores.filter(
    (marcador): marcador is MarcadorSimpleDeDocx =>
      marcador.tipo === 'simple' && (marcador.instruccion ?? '').trim() === '',
  )
}

export function sePuedenProponerInstrucciones(marcadores: readonly MarcadorDeDocx[]): boolean {
  return hayBackend() && camposSinInstruccion(marcadores).length > 0
}

export async function proponerInstrucciones(
  nombreDePlantilla: string,
  marcadores: readonly MarcadorDeDocx[],
): Promise<ResultadoDeConsulta<readonly CampoPropuesto[]>> {
  const campos = camposSinInstruccion(marcadores).map((marcador) => ({
    id: marcador.id,
    nombre: nombreDeMarcador(marcador.textoOriginal),
    contexto: marcador.contexto,
  }))

  const resultado = await pedirAlBackend<{ campos: readonly CampoPropuesto[] }>(
    '/plantillas/instrucciones',
    { nombre: nombreDePlantilla, campos },
  )

  return resultado.ok ? { ok: true, datos: resultado.datos.campos } : resultado
}

/** Aplica lo propuesto sobre los marcadores, sin tocar los que ya tenían instrucción. */
export function conInstruccionesPropuestas(
  marcadores: readonly MarcadorDeDocx[],
  propuestas: readonly CampoPropuesto[],
): readonly MarcadorDeDocx[] {
  const porId = new Map(propuestas.map((propuesta) => [propuesta.id, propuesta]))

  return marcadores.map((marcador) => {
    const propuesta = marcador.tipo === 'simple' ? porId.get(marcador.id) : undefined

    if (propuesta === undefined || marcador.tipo !== 'simple' || (marcador.instruccion ?? '').trim() !== '') {
      return marcador
    }

    return {
      ...marcador,
      instruccion: propuesta.instruccion,
      formato: propuesta.formato,
      modo: propuesta.modo,
      extension: propuesta.extension,
    }
  })
}

File: backend/bitacora/plantillas/instrucciones.py
"""
Proponer, de una vez, qué debe escribir la IA en cada campo de una plantilla.

Configurar una plantilla de quince campos era escribir quince instrucciones a
mano, y ahí es donde se abandonaba: los campos sin instrucción caen en manos
del nombre del marcador, que casi nunca basta ("Contexto", "Detalle"). El
documento ya dice casi todo lo que hace falta —cómo se llama el campo, qué
hay escrito a su alrededor, si después va una lista o un párrafo—, así que la
propuesta se saca de ahí en una sola llamada.

Lo que se propone es EDITABLE y nunca pisa lo ya escrito: quien configuró un
campo a mano tenía una razón, y esto es un punto de partida, no una
corrección.

Vive aparte del endpoint porque es la misma capacidad que va a necesitar el
chat cuando exista ("configúrame esta plantilla"): entonces será una
herramienta más del agente, no una segunda implementación.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any, Protocol, Sequence

from bitacora.compartido.ia import (
    ClienteDeOpenAI,
    TEMPERATURA_DETERMINISTA,
    contenido_del_mensaje,
    leer_json_del_modelo,
    lista_bajo,
    traducir_fallo,
)

FORMATOS = ("parrafo", "lista_vinetas", "lista_numerada")
MODOS = ("redactar", "cita")
EXTENSIONES = ("breve", "media", "extensa")


@dataclass(frozen=True)
class CampoPorConfigurar:
    id: str
    """El nombre tal como se escribió en Word, sin los corchetes: "Resumen de la tesis"."""
    nombre: str
    """El texto del documento alrededor del campo: el rótulo que lo precede y lo que sigue."""
    contexto: str = ""


@dataclass(frozen=True)
class ConfiguracionPropuesta:
    id: str
    instruccion: str
    formato: str
    modo: str
    extension: str


INSTRUCCION = """\
Configuras una plantilla de Word con la que se generan memorias de \
conferencias. Cada campo del documento es un hueco que otra IA rellenará \
después leyendo las fichas de una charla: las ideas citables que dejó.

Recibes los campos con su `nombre` y el `contexto` (lo que está escrito a su \
alrededor en el documento). Para cada uno devuelves cómo debe rellenarse.

Reglas:

1. `instruccion` es una orden en español para quien redacta, de una o dos \
frases, concreta y sobre el contenido de la charla: "Resume la tesis \
principal que defendió el ponente y el argumento con que la sostuvo". No \
escribas el contenido, escribe qué contenido va ahí.
2. Guíate por el nombre y por el contexto. Un campo "Cifras" después de un \
rótulo "Datos de impacto:" pide cifras dichas en la charla; uno llamado \
"Cita" pide una cita.
3. `modo` es `cita` SOLO si el campo pide las palabras exactas de alguien \
(nombres como "Cita", "Frase destacada", "Textual"). En cualquier otro caso, \
`redactar`.
4. `formato` es `lista_vinetas` o `lista_numerada` si el contexto sugiere una \
enumeración (el campo va después de dos puntos y una lista, o el nombre está \
en plural y pide varios elementos). Si no, `parrafo`.
5. `extension`: `breve` para un dato o una frase (fecha, lugar, título, \
cita), `media` para un párrafo, `extensa` solo cuando el campo es el cuerpo \
principal del documento (un resumen general, unas conclusiones amplias).
6. Devuelve una entrada por cada campo que recibiste, con su `id` exacto.

Devuelves un objeto JSON con una única clave `campos`, cuyo valor es una \
lista de objetos \
{"id": "...", "instruccion": "...", "formato": "...", "modo": "...", \
"extension": "..."}."""


class Proponente(Protocol):
    def __call__(
        self, nombre_de_plantilla: str, campos: Sequence[CampoPorConfigurar]
    ) -> tuple[ConfiguracionPropuesta, ...]: ...


def _uno_de(valor: Any, admitidos: tuple[str, ...], por_defecto: str) -> str:
    """Lo que el modelo invente fuera del catálogo cae al valor por defecto, no rompe."""
    return valor if isinstance(valor, str) and valor in admitidos else por_defecto


def proponente_de(cliente: ClienteDeOpenAI, modelo: str) -> Proponente:
    def proponer(
        nombre_de_plantilla: str, campos: Sequence[CampoPorConfigurar]
    ) -> tuple[ConfiguracionPropuesta, ...]:
        entrada = {
            "plantilla": nombre_de_plantilla,
            "campos": [
                {"id": campo.id, "nombre": campo.nombre, "contexto": campo.contexto}
                for campo in campos
            ],
        }

        try:
            respuesta: Any = cliente.chat.completions.create(
                model=modelo,
                temperature=TEMPERATURA_DETERMINISTA,
                response_format={"type": "json_object"},
                messages=[
                    {"role": "system", "content": INSTRUCCION},
                    {"role": "user", "content": json.dumps(entrada, ensure_ascii=False)},
                ],
            )
        except Exception as fallo:  # noqa: BLE001
            raise traducir_fallo(fallo) from fallo

        crudas = lista_bajo(leer_json_del_modelo(contenido_del_mensaje(respuesta)), "campos")

        """
        Emparejado por `id`, igual que la condensación y la redacción: si el
        modelo reordena o se salta uno, hacerlo por posición le pondría a un
        campo la instrucción de otro, y eso no se nota hasta generar la
        memoria. Un campo que no vuelva se queda sin propuesta.
        """
        por_id: dict[str, dict[str, Any]] = {}

        for cruda in crudas:
            if isinstance(cruda, dict) and isinstance(cruda.get("id"), str):
                por_id[cruda["id"]] = cruda

        propuestas: list[ConfiguracionPropuesta] = []

        for campo in campos:
            cruda = por_id.get(campo.id)
            if cruda is None:
                continue

            instruccion = cruda.get("instruccion")
            if not isinstance(instruccion, str) or instruccion.strip() == "":
                continue

            propuestas.append(
                ConfiguracionPropuesta(
                    id=campo.id,
                    instruccion=instruccion.strip(),
                    formato=_uno_de(cruda.get("formato"), FORMATOS, "parrafo"),
                    modo=_uno_de(cruda.get("modo"), MODOS, "redactar"),
                    extension=_uno_de(cruda.get("extension"), EXTENSIONES, "media"),
                )
            )

        return tuple(propuestas)

    return proponer
